LPSPI optimizations 5/19/25

decsny opened the pull request:
Update to optimize the LPSPI configuration latency and code size, as well as fix a few bugs. This is the revival of #85010 plus a few other things.

Changelist:
* Removes call to SDK functions. More info on that below
* Add a wait time limit for the wait_tx_fifo_empty to avoid infinite loop in ISR
* Add major version ID to driver data
* Remove the check for already configured that was preventing register write due to clocking being done in SDK call, now we clocked the LPSPI in the zephyr driver init already, so this was no longer necessary. Also remove a redundant CR write.
* Add clarifying comments to common file for future developers.
* Removes a useless pointer from the driver data to save 4 bytes of RAM per instance.

Removing SDK calls in common file results:
The measurement of 3.7 in #85010 is obviously still valid, but most of the other stuff from that PR already merged, and the comparison from the theoretical 4.1 to this PR is very close, maybe only 1 or 2 hundred bytes more now due to many of the other PRs that have merged over the last few months. The only thing left was removing this LPSPI_MasterInit call in the configure function, so I will hone in on that, which is also very important to the latency of transfers as well. 

All measurements taken on same platform (MCXN947) with same configuration (the one in the spi loopback test). The exact numbers will vary slightly based on platform and configuration (just as before, but should be better in all cases now)
```
Before this PR:
spi_mcux_configure: 900 bytes
fsl_lpspi.c: 1020 bytes
Total: **1920 bytes**
Configuration took 4942 clock cycles

After this PR:
lpspi_configure: 604 bytes
fsl_lpspi: 0 bytes (removed from build)
Total: **604 bytes**
Configuration took 2070 clock cycles
```

Code size improved overall by almost 2x for configure code from around 2000 to 1300 bytes. Or you could consider it to be almost 3x since the new code added is about 400 bytes in the zephyr driver while we remove the 1KB of the SDK doing the same. This is mostly due to SDK having generic code for taking parameters to configure LPSPI in many ways, and having shim code needed to call on this API. With native zephyr driver, we know already for the most part which configuration we need, so we can hardcode it. And the parts we don't know either come from DT (close to hardcoded, just need to read a struct member), or from the spi_config in the API (just needing to decode a bitmask from a struct member). So there removes a lot of the overhead to use a two layer driver. 

Latency improved for configure (and the configure step is the biggest part of software to hardware transfer start latency in the driver) by also about 2x, from almost 5000 to close to 2500 clock cycles. This is mostly due to switching to use a binary search algorithm for the clock divs instead of a linear search. So this improvement is actually algorithmic going from (O(n)) to O(log(n)) depending on which div is optimal. I have suggested to the MCUX SDK team to also make a similar change on that codebase. But I will admit that this is the type of function which would benefit from a direct unit test, which I do not know if that is possible with zephyr. The algorithm is more complex.

Fixes #91015

Commit:
spi_nxp_lpspi: Remove dev pointer from data struct

The dev pointer in the data struct is not being used, so remove it.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
spi_nxp_lpspi: Clarify configuration function

Clarify at the top of the common lpspi file what is the purpose of the
file to be clear to future developers that this file is not supposed to
make any assumption about a particular implementation of the zephyr API
using the LPSPI, because I imagine it could be very likely that more
lpspi implementation will be done in the future to make different
tradeoffs than the current two. Also the current two are different
enough that we should avoid making assumptions even if they currently
hold for both because they might not always, as things change.

We should disable interrupt events while configuring the LPSPI
regardless of implementation. The specific implementation should enable
the interrupts it needs on its own transceive implementation.

Also clarify and simplify some code in the configure function. Namely,
we no longer need to check if we are already configured to write to
registers because a recent commit made it so that we clock the
peripheral from the zephyr driver init instead of upon the MasterInit
call on the SDK. There is also a redundant CR write which I have removed.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
spi_nxp_lpspi: Remove call to MasterInit

For optimization purpose, remove calls to SDK. Since we know exactly
what we want, this results in smaller code size.

Also, this code calculates the SCK parameters more efficiently than the
SDK driver did it by using a binary (instead of linear) search.

Lastly, remove call to LPSPI_Reset in the init call and replace with
native driver code, and remove inclusion of SDK header.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
spi_nxp_lpspi: Add version ID to data struct

Since I expect that the drivers will need to read this version ID maybe
multiple times, instead of repeatedly doing so over the peripheral bus,
it is probably worth it to store a byte in RAM representing this
version. The behavior of the LPSPI is fairly significantly different
between versions. Not enough to warrant separate drivers but enough to
need a few workarounds or different code branches depending on this.

Also, the interrupt based driver is currently using a wrong macro to
read this, and that is a bug.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
spi_nxp_lpspi: Add maximum wait time of fifo empty

Instead of waiting forever and potentially allowing infinite loop on
ISR, wait some arbitrary amount of cycles to error out if it isn't
happening. Still make this configurable for debugging purposes.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
spi_nxp_lpspi: Remove MCUX branding

Since the LPSPI drivers no long use MCUX at all, remove the MCUX
branding, to avoid confusion. In the future if an implementation uses
the MCUX SDK driver, it should specifically be called by MCUX in the
name.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

tbursztyka commented on drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_common.c:
```diff
@@ -116,6 +118,156 @@ static inline int lpspi_validate_xfer_args(const struct spi_config *spi_cfg)
+
+		/* maximum freq won't get better than what we got with previous prescaler */
+		if (clock_freq / (TWO_EXP(prescaler) * 2) < best_freq) {
+			goto done;
```
break?

tbursztyka reviewed (dismissed):
lgtm

decsny commented:
@hakehuang can you help test this one

hakehuang commented:
> @hakehuang can you help test this one

sure, let me kick a test and feedback

pdgendt commented on doc/releases/migration-guide-4.2.rst:
```diff
@@ -500,6 +500,9 @@ OpenThread
 SPI
 ===
 
+* Renamed Renamed :kconfig:option:`CONFIG_SPI_MCUX_LPSPI` to :kconfig:option:`CONFIG_SPI_NXP_LPSPI`,
+  and similar for any child configs for that driver, including
+  :kconfig:option:`CONFIG_SPI_NXP_LPSPI_DMA` and :kconfig:option:`CONFIG_SPI_NXP_LPSPI_CPU`.
```
Symbols that no longer exist can't use the `:kconfig:option:` directive.

```suggestion
* Renamed ``CONFIG_SPI_MCUX_LPSPI`` to :kconfig:option:`CONFIG_SPI_NXP_LPSPI`,
  and similar for any child configs for that driver, including
  :kconfig:option:`CONFIG_SPI_NXP_LPSPI_DMA` and :kconfig:option:`CONFIG_SPI_NXP_LPSPI_CPU`.
```

decsny replied:
I was wondering if it would show up like a red link or something and if that was desired

pdgendt commented on drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_common.c:
```diff
@@ -101,16 +118,166 @@ static inline int lpspi_validate_xfer_args(const struct spi_config *spi_cfg)
+	uint8_t best_scaler = 0, scaler = 0;
+	uint32_t diff, min_diff = 0xFFFFFFFF;
+
+	while (scaler < 256 && min_diff != 0) {
```
`scaler` is a `uint8_t`, this condition does not make sense.

dleach02 replied:
I'm surprised our new sonar static analysis didn't catch this

pdgendt replied:
Should we check if Copilot would? 👀

decsny replied:
woops

teburd replied:
> Should we check if Copilot would? 👀

Coverity would absolutely see this.

Copilot would helplessly suggest a typo fix while burning a small forest to do so.

teburd replied:
It turns out even clang would see this which should be entirely expected?

https://clang-tidy.godbolt.org/z/MWPx148xK

pdgendt replied:
Clang-tidy requires an actual build AFAIK, so not "free"

decsny replied:
I ran the AI below because I was curious what feedback it would give on my PR generally, and it didn't find this. So at least this copilot seems not reliable for rigorous static analysis right now.

decsny commented on drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_common.c:
```diff
@@ -101,16 +118,166 @@ static inline int lpspi_validate_xfer_args(const struct spi_config *spi_cfg)
+	uint32_t diff, min_diff = 0xFFFFFFFF;
+
+	while (scaler < 256 && min_diff != 0) {
+		real_delay *= ((uint64_t)scaler + 1 + (uint64_t)additional_scaler);
+
+		/* Delay must not be less than desired */
+		if (real_delay >= desired_delay_ns) {
+			diff = (uint32_t)(real_delay - (uint64_t)desired_delay_ns);
```
I'll have to look into this more closely, I don't understand this suggestion immediately

decsny commented on drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_common.c:
```diff
@@ -101,16 +118,166 @@ static inline int lpspi_validate_xfer_args(const struct spi_config *spi_cfg)
+			goto done;
+		}
+
+		while (test_div > 0) {
```
um, I already did, so I don't know what it wants here

kartben commented on drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_common.c:
```diff
@@ -126,61 +293,47 @@ int spi_mcux_configure(const struct device *dev, const struct spi_config *spi_cf
+	base->TCR = LPSPI_TCR_CPOL(!!(spi_cfg->operation & SPI_MODE_CPOL)) |
+		    LPSPI_TCR_CPHA(!!(spi_cfg->operation & SPI_MODE_CPHA)) |
+		    LPSPI_TCR_LSBF(!!(spi_cfg->operation & SPI_TRANSFER_LSB)) |
+		    LPSPI_TCR_FRAMESZ(SPI_WORD_SIZE_GET(spi_cfg->operation)) |
```
The FRAMESZ register field requires "number_of_bits - 1", plus, you probably want to use the clamped `word_size` from line 297 (which you probably wanted to as it's actually unused right now?)

decsny replied:
man I hate that I messed that up lol

kartben commented on drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_common.c:
```diff
@@ -126,61 +293,47 @@ int spi_mcux_configure(const struct device *dev, const struct spi_config *spi_cf
+		    LPSPI_TCR_FRAMESZ(SPI_WORD_SIZE_GET(spi_cfg->operation)) |
+		    LPSPI_TCR_PRESCALE(prescaler) | LPSPI_TCR_PCS(spi_cfg->slave);
+
+	lpspi_wait_tx_fifo_empty(dev);
```
check error code?

decsny replied:
true

kartben commented on doc/releases/migration-guide-4.2.rst:
```diff
@@ -500,6 +500,9 @@ OpenThread
 SPI
 ===
 
+* Renamed Renamed :kconfig:option:`CONFIG_SPI_MCUX_LPSPI` to :kconfig:option:`CONFIG_SPI_NXP_LPSPI`,
```
```suggestion
* Renamed :kconfig:option:`CONFIG_SPI_MCUX_LPSPI` to :kconfig:option:`CONFIG_SPI_NXP_LPSPI`,
```

decsny replied:
no, this is valid grammar, it's saying it's renaming the previously renamed config to another name. (just kidding, its a typo)

kartben replied:
> no, this is valid grammar, it's saying it's renaming the previously renamed config to another name. (just kidding, its a typo)

ROFL - you actually got me there :)

kartben reviewed (commented):
so here's for my 100% AI-assisted code review (Gemini 2.5, in this case). I would be really curious to know if the comments related to LPSPI_TCR_FRAMESZ are actually correct (but I think they are) since obviously I have am pretty much clueless when it comes to LPSPI :) And also curious about what @decsny  and @teburd think, of course.

One more comment which I also don't know enough about the code to qualify as being legit or not

> `spi_nxp_lpspi_common.c` - Flawed logic in `lpspi_set_delays`:
> 
> * The conditions if (config->pcs_sck_delay <= minimum_delay_ns) (and similar for SCK-PCS and transfer delays) are counter-intuitive. They cause delay scalers to be configured only if the requested delay is extremely short (e.g., less than or equal to one tick of the prescaled clock). If a longer delay is requested, the if condition is false, and no scaler is set (defaulting to scaler 0, the shortest delay).
> * This is likely the opposite of the intended behavior. The delay scalers are meant to achieve specific, often longer, delays. These if conditions should probably be removed, allowing lpspi_calc_delay to always determine the best scaler (0-255) based on the config->*_delay_ns values. If a delay value in config is 0, lpspi_calc_delay will correctly return scaler = 0.
>

decsny commented:
> One more comment which I also don't know enough about the code to qualify as being legit or not
> 
> > `spi_nxp_lpspi_common.c` - Flawed logic in `lpspi_set_delays`:
> > 
> > * The conditions if (config->pcs_sck_delay <= minimum_delay_ns) (and similar for SCK-PCS and transfer delays) are counter-intuitive. They cause delay scalers to be configured only if the requested delay is extremely short (e.g., less than or equal to one tick of the prescaled clock). If a longer delay is requested, the if condition is false, and no scaler is set (defaulting to scaler 0, the shortest delay).
> > * This is likely the opposite of the intended behavior. The delay scalers are meant to achieve specific, often longer, delays. These if conditions should probably be removed, allowing lpspi_calc_delay to always determine the best scaler (0-255) based on the config->*_delay_ns values. If a delay value in config is 0, lpspi_calc_delay will correctly return scaler = 0.

I had to go back and see how the hardware works and what this code meant since I wrote it months ago now, but, the AI is wrong here. The dividers were already calculated at this point, and what these functions do is actually directly calculate the delay cycles, which will then be divided. So actually the code is doing the correct thing by skipping the calculation if we already know the best minimum is 1 cycle (hardware default, value 0, which will be written at the end of the routine).

decsny commented:
update: fixed the sonarqubecloud complaint and fixed the twister failure

decsny commented on drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_common.c:
```diff
@@ -53,12 +62,22 @@ static inline clock_ip_name_t lpspi_get_clock(LPSPI_Type *const base)
+	while (FIELD_GET(LPSPI_FSR_TXCOUNT_MASK, base->FSR) != 0) {
+		if (!limit_wait) {
+			continue;
+		} else if (arbitrary_cycle_limit-- < 0) {
```
it really doesn't matter

kartben replied:
right but it is, however, a coding guideline violation ([rule 78](https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#:~:text=A%20full%20expression%20containing%20an%20increment%20(%2B%2B)%20or%20decrement%20(%E2%80%93)%20operator%20should%20have%20no%20other%20potential%20side%20effects%20other%20than%20that%20caused%20by%20the%20increment%20or%20decrement%20operator))

kartben replied:
unresolving the comment so that it shows up and gets a chance to be addressed, even if the fix that's needed is different from what copilot initially pointed out :)

decsny replied:
How is it noncompliant? The expression doesn't have any other side effects, just like the rule says

kartben replied:
I might have misinterpreted the rule - I thought it was meant to discourage this kind of construct but looking at all the examples I could find I am probably wrong :) Sorry for the noise.

decsny replied:
@kartben To be clear, a side effect is just anything that is an observable change in the program state besides the evaluation (return value, if you will) of an expression. In this case the expression is "returning" (evaluating) to true or false, and there is one side effect (the decrement of the variable). 

Interestingly, an expression like `x = 10` is also an expression with 1 side effect. The overall expression evaluates to 10, but it also has a "side effect" of changing the value of `x` to 10.

decsny commented on drivers/spi/spi_nxp_lpspi/Kconfig:
```diff
@@ -27,13 +27,29 @@ config SPI_MCUX_LPSPI_DMA
+	  The drivers tend to need to wait on confirming the transmit command
+	  is consumed by the hardware by checking of the TX fifo is emptied.
+	  Usually this should not be a problem and should happen immediately,
+	  but in the rare case of some bug, this option is available to avoid
```
Basically, if for some reason the TX FIFO is not emptying then the command will not be consumed because the SPI hardware configuration commands go in the same FIFO as the TX data. That's one of the things that makes this LPSPI somewhat odd to deal with. Obviously this is not the intended situation and the TX fifo is supposed to be emptied automatically by the hardware for as long as the transfer is going on. But the possibility of an infinite loop in ISR because of some unforeseen programming error which ends the transfer too early or something might scare some people, which is why I am adding this default timeout

decsny commented on drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_common.c:
```diff
@@ -101,17 +120,204 @@ static inline int lpspi_validate_xfer_args(const struct spi_config *spi_cfg)
 
-int spi_mcux_configure(const struct device *dev, const struct spi_config *spi_cfg)
+static uint8_t lpspi_calc_delay(uint32_t desired_delay_ns, uint32_t best_delay_ns,
+				uint32_t prescaled_clock, uint32_t additional_scaler)
```
additional_scaler is just a parameter to be able to abstract all the code instead of having a separate one for DBT. Since the calculation is +1 for DBT for whatever reason as according to the RM.

and no , there's no build warning about that unused parameter. I don't remember what it was for originally but you're right that it's clearly not used anymore.

Commit:
spi_nxp_lpspi: Simplify tx ctx update

There is no need to update the tx context in interrupt instead of
directly after the fill, this just makes the code more complex. Also,
the spi context header already handled iterating over buffers so we can
remove that code too.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
spi_nxp_lpspi: Fix fifo fill logic

Fixing some issues with the TX fifo fill logic in two places:

First in the normal fill function, it didn't take into account a
situation where the TX fifo is already partially filled. This currently
doesn't cause a problem because the driver is written in a way that the
watermark is always 0 for TDF, but in case the watermark were anything
else it would cause a problem.

Second, when filling the TX fifo with NOPS in order to clock the rest of
the RX in from the bus, the calculation regarding the current TX fifo
length was just wrong and was leading to a bug in some cases where there
was a subtraction underflow and billions of NOPs were being filled.
Also, there could be a problem where a few extra NOPs are put in the TX
fifo if we don't count what we already have in the TX fifo, so fixing
that as well.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

pdgendt commented:
Running this on MIMXRT1170-EVKB:

```
------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [spi_extra_api_features]: pass = 2, fail = 0, skip = 0, total = 2 duration = 0.003 seconds
 - PASS - [spi_extra_api_features.test_spi_hold_on_cs] duration = 0.002 seconds
 - PASS - [spi_extra_api_features.test_spi_lock_release] duration = 0.001 seconds

SUITE FAIL -  84.00% [spi_loopback]: pass = 21, fail = 4, skip = 0, total = 25 duration = 0.408 seconds
 - PASS - [spi_loopback.test_nop_nil_bufs] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_async_call] duration = 0.137 seconds
 - PASS - [spi_loopback.test_spi_complete_large_transfers] duration = 0.136 seconds
 - PASS - [spi_loopback.test_spi_complete_loop_mode_0] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_complete_loop_mode_1] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_complete_loop_mode_2] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_complete_loop_mode_3] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_complete_multiple] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_complete_multiple_timed] duration = 0.008 seconds
 - FAIL - [spi_loopback.test_spi_concurrent_transfer_different_spec] duration = 0.032 seconds
 - FAIL - [spi_loopback.test_spi_concurrent_transfer_same_spec] duration = 0.032 seconds
 - PASS - [spi_loopback.test_spi_null_rx_buf_set] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_null_tx_buf] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_null_tx_buf_set] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_rx_bigger_than_tx] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_rx_every_4] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_rx_half_end] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_rx_half_start] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_same_buf_cmd] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_word_size_16] duration = 0.001 seconds
 - FAIL - [spi_loopback.test_spi_word_size_24] duration = 0.022 seconds
 - PASS - [spi_loopback.test_spi_word_size_32] duration = 0.004 seconds
 - PASS - [spi_loopback.test_spi_word_size_7] duration = 0.001 seconds
 - FAIL - [spi_loopback.test_spi_word_size_9] duration = 0.020 seconds
 - PASS - [spi_loopback.test_spi_write_back] duration = 0.001 seconds

------ TESTSUITE SUMMARY END ------
```

MikhailSiomin reviewed (commented):
https://github.com/zephyrproject-rtos/zephyr/blob/58c0fea675c003da7ec96568246d308564bf47fa/drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi.c#L167-L168
`current_buf->len` stores the number of bytes, but `buf_len` argumet of `lpspi_fill_tx_fifo` function multiply by the number of bytes in the word.

decsny commented on drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_common.c:
```diff
@@ -101,16 +118,166 @@ static inline int lpspi_validate_xfer_args(const struct spi_config *spi_cfg)
+			}
+
+			/* if our best found is a match, we're done */
+			if (best_freq == desired_freq) {
```
I think you are reviewing an old revision, I fixed this algorithm a lot and tested millions of combinations, and had no problems

MikhailSiomin replied:
@decnsy, sorry, I created a comment incorrectly, which pulled up an old comment. I wanted to report about the `lpspi_fill_tx_fifo` function

MikhailSiomin commented:
@decsny, unfortunately, now I can't get it to work with nand-flash memory and some sensors, unlike the driver of zephyr-4.0.0.

decsny commented:
> @decsny, unfortunately, now I can't get it to work with nand-flash memory and some sensors, unlike the driver of zephyr-4.0.0.

Are you saying this PR is breaking it for you? Or are you saying it's an issue on main? If it's an issue on main then please submit a bug report about it with more details

MikhailSiomin commented:
> > @decsny, unfortunately, now I can't get it to work with nand-flash memory and some sensors, unlike the driver of zephyr-4.0.0.
> 
> Are you saying this PR is breaking it for you? Or are you saying it's an issue on main? If it's an issue on main then please submit a bug report about it with more details

main and this PR do not work for me. I will try to figure out this problem and give feedback in the issue if the problem is on the driver side. Thanks

Commit:
spi_nxp_lpspi: Fix unit of buf_len in fill_tx_fifo

The buf_len parameter of lpspi_fill_tx_fifo is supposed to be bytes, so
we do not need to convert it. This could cause an issue if the end of
the buffer is less bytes than the word size.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

igalloway commented:
Thanks for the great effort so far. 
We need this PR to be finished to a reasonable level so it can be pushed - so we can then continue to iterate on the remaining associated issues.
Could you please allocate some time to work on it to get it in decent shape?
Currently it is blocking proper SPI performance needed for high data rate sensors with low latency.
We would very much appreciate increasing the priority somehow.

decsny commented:
update:

- rebase
- added fix suggested by @MikhailSiomin regarding filling tx fifo.  There could have been an issue if the very end of a buffer was smaller than the size of a word, in other words, buf_size % word_size != 0
- confirmed I made a mistake on the last push by calculating the prescaler after the delay scalars (the prescaler must be known for the delay calculations), thanks @smaherdev for reporting an issue with the delays on this PR so I realized this. I had the DNM because of this issue, so now that I think I fixed it, I removed the DNM.

So now AFAIK the PR is ready

hakehuang commented:
with my mimxrt1170_evk@B board
```
*** Booting Zephyr OS build v4.1.0-5591-g28f940aa4d4c ***
SPI test on buffers TX/RX 0x80004160/0x80004140, frame size = 8, DMA enabled
Polling...Running TESTSUITE spi_extra_api_features
===================================================================
START - test_spi_hold_on_cs
 PASS - test_spi_hold_on_cs in 0.002 seconds
===================================================================
START - test_spi_lock_release
 PASS - test_spi_lock_release in 0.001 seconds
===================================================================
TESTSUITE spi_extra_api_features succeeded
Running TESTSUITE spi_extra_api_features
===================================================================
START - test_spi_hold_on_cs
 PASS - test_spi_hold_on_cs in 0.002 seconds
===================================================================
START - test_spi_lock_release
 PASS - test_spi_lock_release in 0.001 seconds
===================================================================
TESTSUITE spi_extra_api_features succeeded
Running TESTSUITE spi_loopback
===================================================================
Testing loopback spec: SLOW
START - test_nop_nil_bufs
 PASS - test_nop_nil_bufs in 0.001 seconds
===================================================================
START - test_spi_async_call
 PASS - test_spi_async_call in 0.138 seconds
===================================================================
START - test_spi_complete_large_transfers
 PASS - test_spi_complete_large_transfers in 0.136 seconds
===================================================================
START - test_spi_complete_loop_mode_0
 PASS - test_spi_complete_loop_mode_0 in 0.001 seconds
===================================================================
START - test_spi_complete_loop_mode_1
 PASS - test_spi_complete_loop_mode_1 in 0.001 seconds
===================================================================
START - test_spi_complete_loop_mode_2
 PASS - test_spi_complete_loop_mode_2 in 0.001 seconds
===================================================================
START - test_spi_complete_loop_mode_3
 PASS - test_spi_complete_loop_mode_3 in 0.001 seconds
===================================================================
START - test_spi_complete_multiple
 PASS - test_spi_complete_multiple in 0.002 seconds
===================================================================
START - test_spi_complete_multiple_timed
Transfer took 959 us vs theoretical minimum 864 us
Latency measurement: 95 us
 PASS - test_spi_complete_multiple_timed in 0.008 seconds
===================================================================
START - test_spi_concurrent_transfer_different_spec
SPI concurrent transfer data mismatch, spec 0x80008bfc
SPI concurrent transfer data mismatch, spec 0x80008bcc
SPI concurrent transfer data mismatch, spec 0x80008be4

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/spi/spi_loopback/src/spi.c:787: test_spi_concurrent_transfer_helper: atomic_get(&thread_test_fails) not equal to 0

 FAIL - test_spi_concurrent_transfer_different_spec in 0.032 seconds
===================================================================
START - test_spi_concurrent_transfer_same_spec
SPI concurrent transfer data mismatch, spec 0x80008028
SPI concurrent transfer data mismatch, spec 0x80008028
SPI concurrent transfer data mismatch, spec 0x80008028

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/spi/spi_loopback/src/spi.c:787: test_spi_concurrent_transfer_helper: atomic_get(&thread_test_fails) not equal to 0

 FAIL - test_spi_concurrent_transfer_same_spec in 0.032 seconds
===================================================================
START - test_spi_null_rx_buf_set
 PASS - test_spi_null_rx_buf_set in 0.001 seconds
===================================================================
START - test_spi_null_tx_buf
 PASS - test_spi_null_tx_buf in 0.001 seconds
===================================================================
START - test_spi_null_tx_buf_set
 PASS - test_spi_null_tx_buf_set in 0.001 seconds
===================================================================
START - test_spi_rx_bigger_than_tx
 PASS - test_spi_rx_bigger_than_tx in 0.001 seconds
===================================================================
START - test_spi_rx_every_4
 PASS - test_spi_rx_every_4 in 0.001 seconds
===================================================================
START - test_spi_rx_half_end
 PASS - test_spi_rx_half_end in 0.001 seconds
===================================================================
START - test_spi_rx_half_start
 PASS - test_spi_rx_half_start in 0.001 seconds
===================================================================
START - test_spi_same_buf_cmd
 PASS - test_spi_same_buf_cmd in 0.001 seconds
===================================================================
START - test_spi_word_size_16
 PASS - test_spi_word_size_16 in 0.001 seconds
===================================================================
START - test_spi_word_size_24

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/spi/spi_loopback/src/spi.c:643: spi_loopback_test_word_size: (memcmp(compare_data, rx_buffer, buffer_size) is true)
24-bit word buffer contents are different
 FAIL - test_spi_word_size_24 in 0.022 seconds
===================================================================
START - test_spi_word_size_32
 PASS - test_spi_word_size_32 in 0.004 seconds
===================================================================
START - test_spi_word_size_7
 PASS - test_spi_word_size_7 in 0.001 seconds
===================================================================
START - test_spi_word_size_9

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/spi/spi_loopback/src/spi.c:643: spi_loopback_test_word_size: (memcmp(compare_data, rx_buffer, buffer_size) is true)
9-bit word buffer contents are different
 FAIL - test_spi_word_size_9 in 0.020 seconds
===================================================================
START - test_spi_write_back
 PASS - test_spi_write_back in 0.001 seconds
===================================================================
TESTSUITE spi_loopback failed.
Running TESTSUITE spi_loopback
===================================================================
Testing loopback spec: FAST
START - test_nop_nil_bufs
 PASS - test_nop_nil_bufs in 0.001 seconds
===================================================================
START - test_spi_async_call
 PASS - test_spi_async_call in 0.010 seconds
===================================================================
START - test_spi_complete_large_transfers
 PASS - test_spi_complete_large_transfers in 0.009 seconds
===================================================================
START - test_spi_complete_loop_mode_0
 PASS - test_spi_complete_loop_mode_0 in 0.001 seconds
===================================================================
START - test_spi_complete_loop_mode_1
 PASS - test_spi_complete_loop_mode_1 in 0.001 seconds
===================================================================
START - test_spi_complete_loop_mode_2
 PASS - test_spi_complete_loop_mode_2 in 0.001 seconds
===================================================================
START - test_spi_complete_loop_mode_3
 PASS - test_spi_complete_loop_mode_3 in 0.001 seconds
===================================================================
START - test_spi_complete_multiple
 PASS - test_spi_complete_multiple in 0.001 seconds
===================================================================
START - test_spi_complete_multiple_timed
Transfer took 106 us vs theoretical minimum 27 us
Latency measurement: 79 us
 PASS - test_spi_complete_multiple_timed in 0.007 seconds
===================================================================
START - test_spi_concurrent_transfer_different_spec
SPI concurrent transfer data mismatch, spec 0x80008bfc
SPI concurrent transfer data mismatch, spec 0x80008bcc
SPI concurrent transfer data mismatch, spec 0x80008be4

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/spi/spi_loopback/src/spi.c:787: test_spi_concurrent_transfer_helper: atomic_get(&thread_test_fails) not equal to 0

 FAIL - test_spi_concurrent_transfer_different_spec in 0.031 seconds
===================================================================
START - test_spi_concurrent_transfer_same_spec
SPI concurrent transfer data mismatch, spec 0x80008040
SPI concurrent transfer data mismatch, spec 0x80008040
SPI concurrent transfer data mismatch, spec 0x80008040

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/spi/spi_loopback/src/spi.c:787: test_spi_concurrent_transfer_helper: atomic_get(&thread_test_fails) not equal to 0

 FAIL - test_spi_concurrent_transfer_same_spec in 0.031 seconds
===================================================================
START - test_spi_null_rx_buf_set
 PASS - test_spi_null_rx_buf_set in 0.001 seconds
===================================================================
START - test_spi_null_tx_buf
 PASS - test_spi_null_tx_buf in 0.001 seconds
===================================================================
START - test_spi_null_tx_buf_set
 PASS - test_spi_null_tx_buf_set in 0.001 seconds
===================================================================
START - test_spi_rx_bigger_than_tx
 PASS - test_spi_rx_bigger_than_tx in 0.001 seconds
===================================================================
START - test_spi_rx_every_4
 PASS - test_spi_rx_every_4 in 0.001 seconds
===================================================================
START - test_spi_rx_half_end
 PASS - test_spi_rx_half_end in 0.001 seconds
===================================================================
START - test_spi_rx_half_start
 PASS - test_spi_rx_half_start in 0.001 seconds
===================================================================
START - test_spi_same_buf_cmd
 PASS - test_spi_same_buf_cmd in 0.001 seconds
===================================================================
START - test_spi_word_size_16
 PASS - test_spi_word_size_16 in 0.001 seconds
===================================================================
START - test_spi_word_size_24

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/spi/spi_loopback/src/spi.c:643: spi_loopback_test_word_size: (memcmp(compare_data, rx_buffer, buffer_size) is true)
24-bit word buffer contents are different
 FAIL - test_spi_word_size_24 in 0.020 seconds
===================================================================
START - test_spi_word_size_32
 PASS - test_spi_word_size_32 in 0.001 seconds
===================================================================
START - test_spi_word_size_7
 PASS - test_spi_word_size_7 in 0.001 seconds
===================================================================
START - test_spi_word_size_9

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/spi/spi_loopback/src/spi.c:643: spi_loopback_test_word_size: (memcmp(compare_data, rx_buffer, buffer_size) is true)
9-bit word buffer contents are different
 FAIL - test_spi_word_size_9 in 0.019 seconds
===================================================================
START - test_spi_write_back
 PASS - test_spi_write_back in 0.001 seconds
===================================================================
TESTSUITE spi_loopback failed.

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [spi_extra_api_features]: pass = 2, fail = 0, skip = 0, total = 2 duration = 0.003 seconds
 - PASS - [spi_extra_api_features.test_spi_hold_on_cs] duration = 0.002 seconds
 - PASS - [spi_extra_api_features.test_spi_lock_release] duration = 0.001 seconds

SUITE FAIL -  84.00% [spi_loopback]: pass = 21, fail = 4, skip = 0, total = 25 duration = 0.410 seconds
 - PASS - [spi_loopback.test_nop_nil_bufs] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_async_call] duration = 0.138 seconds
 - PASS - [spi_loopback.test_spi_complete_large_transfers] duration = 0.136 seconds
 - PASS - [spi_loopback.test_spi_complete_loop_mode_0] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_complete_loop_mode_1] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_complete_loop_mode_2] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_complete_loop_mode_3] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_complete_multiple] duration = 0.002 seconds
 - PASS - [spi_loopback.test_spi_complete_multiple_timed] duration = 0.008 seconds
 - FAIL - [spi_loopback.test_spi_concurrent_transfer_different_spec] duration = 0.032 seconds
 - FAIL - [spi_loopback.test_spi_concurrent_transfer_same_spec] duration = 0.032 seconds
 - PASS - [spi_loopback.test_spi_null_rx_buf_set] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_null_tx_buf] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_null_tx_buf_set] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_rx_bigger_than_tx] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_rx_every_4] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_rx_half_end] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_rx_half_start] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_same_buf_cmd] duration = 0.001 seconds
 - PASS - [spi_loopback.test_spi_word_size_16] duration = 0.001 seconds
 - FAIL - [spi_loopback.test_spi_word_size_24] duration = 0.022 seconds
 - PASS - [spi_loopback.test_spi_word_size_32] duration = 0.004 seconds
 - PASS - [spi_loopback.test_spi_word_size_7] duration = 0.001 seconds
 - FAIL - [spi_loopback.test_spi_word_size_9] duration = 0.020 seconds
 - PASS - [spi_loopback.test_spi_write_back] duration = 0.001 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
RunID: 6a393893f19313cd73eab94086cba3a8
PROJECT EXECUTION FAILED

```

decsny commented:
At this point I am not accepting any more comments for scope expansions on this PR, there's been enough, this PR is becoming a kitchen sink.

Reviewers just tell me if this PR is specifically regressing something, or review _only the code being changed on this PR_ for if you don't like some part of it.

If anyone wants to have some other feature or optimization or bug fix added to the driver, please make a separate github issue (minding the ones that already exist) or submit your own PR, otherwise this one is never going to merge.

The only thing I'm left to do right now is apparently the delay scalars are still being regressed on this PR according to @smaherdev so I am going to try to figure that out, so the DNM is back for now

Commit:
spi_nxp_lpspi: Fix extra byte sent on v1 lpspi

This stupid errata will not leave me alone, here is another bandaid to
deal with an issue where an extra byte was being sent on version 1
LPSPIs due to the algorithm of filling NOPs when only RX is left was not
expecting the situation where the LPSPI actually consumed everything
from the fifo but is not sending it due to this ridiculous stalling errata.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

decsny commented:
I found that issue with the delay scalers was because I somewhat mindlessly copied the algorithm from the SDK without thinking too much about it, assuming it would be the same, but I made some mistakes which led to most cases being "edge" cases. 

The code that was there for those scalers on this PR before latest push was atrociously bad due to my mistakes, so I apologize to anyone who tested this PR for wasting their time.

The good news is that now I've fixed it and done a lot of regression testing with many different scaler and clock frequency combinations, and I'm pretty confident about it now. And on top of that, when I was looking into this, I realized that the algorithm used by the SDK was ridiculously overcomplicated, amazingly so. I've done a new algorithm which has the same behavior as the old one but is way simpler and therefore should be much faster. Even though it is still a basic naive linear search, nevertheless, what I measure for the new code size of the configure functions are hundreds of bytes less than what is reported in the original PR message, and the configure latency is hundreds of cycles less as well.

decsny commented:
update:
- rebase
- fixed delay scalers
- removed fill_len from driver data because it is no longer used (forgot to do this before)
- fixed extra byte issue from #90614 (I think)

decsny commented on drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_common.c:
```diff
@@ -101,17 +122,193 @@ static inline int lpspi_validate_xfer_args(const struct spi_config *spi_cfg)
+
+	if (low == high) {
+		new_freq = lpspi_calc_sck_freq(src_clk_hz, high, prescaler);
+		if (new_freq > min_freq && new_freq <= max_freq) {
```
wrong

decsny commented on drivers/spi/spi_nxp_lpspi/Kconfig:
```diff
@@ -27,13 +27,26 @@ config SPI_MCUX_LPSPI_DMA
-endif # SPI_MCUX_LPSPI
+config SPI_NXP_LPSPI_TXFIFO_WAIT_CYCLES
+	int "Number of CPU cycles to wait on TX fifo empty"
+	default 0 if DEBUG
+	default 10000
```
?

hakehuang replied:
@decsny now zephyr project enables copilot review, but copilot is not good at Kconfig staff.

pdgendt commented:
> @pdgendt the issue of the "missing clock pulse" is most likely I think a signal analyzer sampling rate issue, or maybe even a pinctrl issue, or both combined. I'm assuming that since you're using the RT1170 that the function clock frequency of the lpspi is actually extremely fast (probably something like 48 MHz i'm guessing is how you have it configured?) and the default delay is 2 cycles of that, so you would need a sampling rate of probably a few hundred MHz to see it cleanly assuming the pad can even reach the low threshold in that time

The target device is an nRF7002, which has a max clock frequency of 8 MHz, but with the logic analyzer I set it to 1 MHz and sample at 40MS/s, so I don't think the analyzer is the issue.

Also note my previous comment that it works if I use a high enough `transfer-delay`, see (8 MHz):
![image](https://github.com/user-attachments/assets/19391396-4fac-4d35-bcc6-3647c052e901)

decsny commented:
As for your screenshot of the SCKPCS and PCSSCK, I'll see if I can find an RT1170 board and reproduce, that delay of 5 us should have been possible I would think.

decsny commented:
> > @pdgendt the issue of the "missing clock pulse" is most likely I think a signal analyzer sampling rate issue, or maybe even a pinctrl issue, or both combined. I'm assuming that since you're using the RT1170 that the function clock frequency of the lpspi is actually extremely fast (probably something like 48 MHz i'm guessing is how you have it configured?) and the default delay is 2 cycles of that, so you would need a sampling rate of probably a few hundred MHz to see it cleanly assuming the pad can even reach the low threshold in that time
> 
> The target device is an nRF7002, which has a max clock frequency of 8 MHz, but with the logic analyzer I set it to 1 MHz and sample at 40MS/s, so I don't think the analyzer is the issue.
> 
> Also note my previous comment that it works if I use a high enough `transfer-delay`, see (8 MHz)

I'm not sure you're clear on what I'm saying. I did note that you found it "works" if you use a higher `transfer-delay`, which is the main reason that led me to say what I said. I'm saying the fact that the transfer delay is so short is making it look like it didn't happen due to the sampling resolution or could be not reaching low threshold due to pinctrl capability. 

The fact that you configured the max frequency of the spi device to be 1 MHz is irrelevant to what I'm saying. But, I saw from your screenshot that the frequency of your SCK was 1 MHz and you said CCR was written as 0x30, so that's how I calculated that your source clock frequency to LPSPI is probably around 48 MHz. Like I said, it's this source clock frequency, not the SCK frequency, that determines how the delays are calculated. So if transfer-delay is set to the default of 0, the driver will configure the absolute minimum delay, which is 2 cycles of the source clock, i.e. 41 ns, so you probably need triple digits of MS/s to see that pulse.

pdgendt commented on drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_common.c:
```diff
@@ -101,17 +122,180 @@ static inline int lpspi_validate_xfer_args(const struct spi_config *spi_cfg)
+
+	delay_cycles = (uint64_t)prescaled_clock * desired_delay_ns;
+	delay_cycles = DIV_ROUND_UP(delay_cycles, NSEC_PER_SEC);
+	delay_cycles = CLAMP(delay_cycles, min_cycles, max_cycles);
+
+	return delay_cycles - min_cycles;
```
Better safe than sorry?

```suggestion
	delay_cycles = CLAMP(delay_cycles, min_cycles, max_cycles);
	delay_cycles -= min_cycles;

	return MIN(delay_cycles, UINT8_MAX);
```

decsny replied:
`max_cycles` = 255 + `min_cycles`

delay_cycles gets clamped already to high of `max_cycles`, so assuming that is the maximum:

`max_cycles` - 255 = `min_cycles` -> 
`max_cycles` - `min_cycles` = 255
maximum value returned then would be 255

I won't fight too hard against making the change if you still think it's worthwhile but I think it's pretty obviously mathematically proveable that it won't be needed...

However, an argument could be made though that it could make the code above that more robust against someone changing it later erroneously. And it's negligible overhead to add that macro. What do you think

decsny replied:
I have an idea actually to satisfy your concern without adding extra code, I'll push it in a minute

pdgendt commented on drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_common.c:
```diff
@@ -126,62 +326,55 @@ int spi_mcux_configure(const struct device *dev, const struct spi_config *spi_cf
 	}
 
+	/* For the purpose of configuring the LPSPI, 8 is the minimum frame size for the hardware */
+	word_size = word_size < 8 ? 8 : word_size;
```
Not changed yet?

pdgendt reviewed (commented):
Thanks @decsny! The changes do make a lot of sense!

Will test them soon and report the results.

decsny commented:
@pdgendt I pushed a minor change for you to review based on the comments above, but at this time of the day in the week, I didn't test on hardware yet, I'll do it on monday

EmilioCBen reviewed (approved):
LGTM! Great work!

mmahadevan108 commented on drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_common.c:
```diff
@@ -60,12 +60,24 @@ static inline clock_ip_name_t lpspi_get_clock(LPSPI_Type *const base)
+		if (arbitrary_cycle_limit-- < 0) {
+			LOG_WRN("Failed waiting for TX fifo empty");
+			return -EIO;
+		}
```
Why not use `k_busy_wait()` here for maybe 1usec per iteration and use the Kconfig  to determine total the wait time in usec. This way the behavior will not vary by the processor speed.

decsny replied:
It's supposed to vary by processor speed. It's just a number of attempts to try before quitting. It's simply a basic safety mechanism, which is just about being more safe than sorry, not a unique feature. It shouldn't even be needed. Most likely this consumption by the fifo happens immediately, and this loop isn't necessary. But people wanted to expand the scope of this PR, so here it is.  I'm not overcomplicating it any more by making it time -based based on CPU frequency calculations.

Also the specific suggestion to use k_busy_wait is not a good idea. That function is not isr safe. This function is called in isr context. k_busy_wait is implemented by polling kernel time, which can't increment due to this function would be blocking in isr context, causing a likely infinite loop in the kernel.
